File: backend/app/schemas/model_config.py
"""Pydantic schemas for model configuration management."""

from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.common import TimestampMixin


class ModelConfigBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    model_id: str = Field(..., min_length=1, max_length=100)
    context_window: int | None = Field(default=None, ge=0)
    max_output_tokens: int | None = Field(default=None, ge=0)
    pool_group: str | None = Field(default=None, max_length=50)
    default_temperature: float | None = Field(default=None, ge=0, le=2)
    default_top_p: float | None = Field(default=None, ge=0, le=1)
    timeout_sec: int | None = Field(default=None, ge=1, le=600)
    retry_count: int | None = Field(default=None, ge=0, le=10)
    support_text: bool = True
    support_image_input: bool = False
    support_image_output: bool = False
    support_tools: bool = True
    support_stream: bool = True
    support_code: bool = False
    support_long_text: bool = False
    enabled: bool = True
    sort: int = Field(default=0, ge=0)


class ModelConfigCreate(ModelConfigBase):
    provider_id: UUID


class ModelConfigUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    model_id: str | None = Field(default=None, min_length=1, max_length=100)
    context_window: int | None = Field(default=None, ge=0)
    max_output_tokens: int | None = Field(default=None, ge=0)
    pool_group: str | None = Field(default=None, max_length=50)
    default_temperature: float | None = Field(default=None, ge=0, le=2)
    default_top_p: float | None = Field(default=None, ge=0, le=1)
    timeout_sec: int | None = Field(default=None, ge=1, le=600)
    retry_count: int | None = Field(default=None, ge=0, le=10)
    support_text: bool | None = None
    support_image_input: bool | None = None
    support_image_output: bool | None = None
    support_tools: bool | None = None
    support_stream: bool | None = None
    support_code: bool | None = None
    support_long_text: bool | None = None
    enabled: bool | None = None
    sort: int | None = Field(default=None, ge=0)


class ModelConfigInfo(ModelConfigBase, TimestampMixin):
    id: UUID
    provider_id: UUID

    model_config = {"from_attributes": True}


class ModelProviderBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    provider: str = Field(..., min_length=1, max_length=50)
    api_endpoint: str = Field(..., min_length=1, max_length=500)
    protocol_type: str = Field(default="openai-compatible", min_length=1, max_length=50)
    api_version: str | None = Field(default=None, max_length=50)
    remark: str | None = Field(default=None, max_length=500)
    enabled: bool = True
    sort: int = Field(default=0, ge=0)


class ModelProviderCreate(ModelProviderBase):
    api_key: str | None = Field(default=None, max_length=1000)


class ModelProviderUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=100)
    provider: str | None = Field(default=None, min_length=1, max_length=50)
    api_endpoint: str | None = Field(default=None, min_length=1, max_length=500)
    api_key: str | None = Field(default=None, max_length=1000)
    protocol_type: str | None = Field(default=None, min_length=1, max_length=50)
    api_version: str | None = Field(default=None, max_length=50)
    remark: str | None = Field(default=None, max_length=500)
    enabled: bool | None = None
    sort: int | None = Field(default=None, ge=0)


class ModelProviderInfo(ModelProviderBase, TimestampMixin):
    id: UUID
    has_api_key: bool
    models: list[ModelConfigInfo] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class TestConnectionRequest(BaseModel):
    provider_id: UUID
    model_id: str | None = Field(default=None, max_length=100)


class TestConnectionResult(BaseModel):
    reachable: bool
    status_code: int | None = None
    message: str

File: ui/lib/api/model-config.ts
import { apiClient } from "./client";
import type { ApiPaginated, ApiSuccess } from "./auth";

export interface ModelConfigInfo {
  id: string;
  provider_id: string;
  name: string;
  model_id: string;
  context_window: number | null;
  max_output_tokens: number | null;
  pool_group: string | null;
  default_temperature: number | null;
  default_top_p: number | null;
  timeout_sec: number | null;
  retry_count: number | null;
  support_text: boolean;
  support_image_input: boolean;
  support_image_output: boolean;
  support_tools: boolean;
  support_stream: boolean;
  support_code: boolean;
  support_long_text: boolean;
  enabled: boolean;
  sort: number;
  created_at: string;
  updated_at: string | null;
}

export interface ModelProviderInfo {
  id: string;
  name: string;
  provider: string;
  api_endpoint: string;
  protocol_type: string;
  api_version: string | null;
  remark: string | null;
  enabled: boolean;
  sort: number;
  has_api_key: boolean;
  models: ModelConfigInfo[];
  created_at: string;
  updated_at: string | null;
}

export interface ModelProviderPayload {
  name: string;
  provider: string;
  api_endpoint: string;
  api_key?: string;
  protocol_type: string;
  api_version?: string | null;
  remark?: string | null;
  enabled?: boolean;
  sort?: number;
}

export interface ModelConfigPayload {
  provider_id: string;
  name: string;
  model_id: string;
  context_window?: number | null;
  max_output_tokens?: number | null;
  pool_group?: string | null;
  default_temperature?: number | null;
  default_top_p?: number | null;
  timeout_sec?: number | null;
  retry_count?: number | null;
  support_text?: boolean;
  support_image_input?: boolean;
  support_image_output?: boolean;
  support_tools?: boolean;
  support_stream?: boolean;
  support_code?: boolean;
  support_long_text?: boolean;
  enabled?: boolean;
  sort?: number;
}

export interface TestConnectionResult {
  reachable: boolean;
  status_code: number | null;
  message: string;
}

export function listModelProviders(params?: {
  p?: number;
  page_size?: number;
  keyword?: string;
}) {
  return apiClient.get<ApiPaginated<ModelProviderInfo>>("/model-config/providers", { params });
}

export function listAllModelProviders() {
  return apiClient.get<ApiSuccess<ModelProviderInfo[]>>("/model-config/providers/all-with-models");
}

export function createModelProvider(body: ModelProviderPayload) {
  return apiClient.post<ApiSuccess<ModelProviderInfo>>("/model-config/providers", body);
}

export function updateModelProvider(id: string, body: Partial<ModelProviderPayload>) {
  return apiClient.put<ApiSuccess<ModelProviderInfo>>(`/model-config/providers/${id}`, body);
}

export function deleteModelProvider(id: string) {
  return apiClient.delete<{ success: boolean; message: string }>(`/model-config/providers/${id}`);
}

export function testModelProviderConnection(providerId: string, modelId?: string) {
  return apiClient.post<ApiSuccess<TestConnectionResult>>(
    "/model-config/providers/test-connection",
    { provider_id: providerId, model_id: modelId }
  );
}

export function createModelConfig(body: ModelConfigPayload) {
  return apiClient.post<ApiSuccess<ModelConfigInfo>>("/model-config/models", body);
}

export function updateModelConfig(id: string, body: Partial<ModelConfigPayload>) {
  return apiClient.put<ApiSuccess<ModelConfigInfo>>(`/model-config/models/${id}`, body);
}

export function deleteModelConfig(id: string) {
  return apiClient.delete<{ success: boolean; message: string }>(`/model-config/models/${id}`);
}
